File: 0410. Split Array Largest Sum/Solution.py
class Solution(object):
    def splitArray(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        left = max(nums)
        right = sum(nums)

        while left<=right:
            mid = left + (right-left)//2
            curr = self.minSubSum(nums,mid)

            if(curr<=k):
                right = mid-1
            else:
                left = mid+1
        return left

    
    def minSubSum(self,nums,d):
        count = 1
        curr = 0

        for i in nums:
            if (curr+i)<=d:
                curr+=i
            else:
                count+=1
                curr=i
        
        return count 

File: 0410. Split Array Largest Sum/Solution.java
class Solution {
    public int splitArray(int[] nums, int k) {
        int left = Integer.MIN_VALUE;
        int right = 0;
        for(int i : nums){
            left = Math.max(left,i);
            right+=i;
        }

        while(left<=right){
            int mid = left + (right-left)/2;
            int curr = minSubSum(nums,mid);

            if(curr<=k){
                right = mid - 1;
            }
            else{
                left = mid + 1;
            }
        }
        return left;
    }

    private int minSubSum(int[] nums,int d){
        int count = 1;
        int curr = 0;
        for(int i=0;i<nums.length;i++){
            if(curr+nums[i]<=d){
                curr += nums[i];
            }
            else{
                count++;
                curr = nums[i];
            }
        }
        return count;
    }
}